gc_registry/certificate/services.py

```diff
@@ -304,40 +304,199 @@ def process_certificate_action(
+
+    For each GC Bundle returned from the query, this function will split the bundle
+    if the desired GC quantity or percentage is less than the GC Bundle quantity, otherwise
+    it will return the GC Bundle unsplit.
+
+    Args:
+        certificates_from_query (list[GranularCertificateBundle]): The certificates from the query
+        certificate_bundle_action (GranularCertificateAction): The certificate action
+        write_session (Session): The database write session
+        read_session (Session): The database read session
+        esdb_client (EventStoreDBClient): The EventStoreDB client
+
+    Returns:
+        list[GranularCertificateBundle]: The list of certificates to transfer, split if required
+                                         such that the quantity of each bundle is equal to or
+                                         less than the desired bundle quantity, if provided, or
+                                         the percentage of the total certificates in the bundle.
+
+    """
+    # Just return the certificates from the query if no quantity or percentage is provided
+    if (certificate_bundle_action.certificate_quantity is None) & (
+        certificate_bundle_action.certificate_bundle_percentage is None
+    ):
+        return certificates_from_query
+
+    certificates_to_transfer = []
+
+    if certificate_bundle_action.certificate_quantity is not None:
+        certificates_to_split = [
+            certificate_bundle_action.certificate_quantity
+            for i in range(len(certificates_from_query))
+        ]
+    elif certificate_bundle_action.certificate_bundle_percentage is not None:
+        certificates_to_split = [
+            int(
+                certificate_bundle_action.certificate_bundle_percentage
+                * certificate_from_query.bundle_quantity
+                / 100
+            )
+            for certificate_from_query in certificates_from_query
+        ]
+
+    # Only check the bundle quantity if the query on bundle quantity parameter is provided,
+    # otherwise, split the bundle based on the percentage of the total certificates in the bundle
+    for idx, gc_bundle in enumerate(certificates_from_query):
+        gc_bundle = write_session.merge(gc_bundle)
+
+        if certificate_bundle_action.certificate_quantity is not None:
+            if (
+                gc_bundle.bundle_quantity
+                <= certificate_bundle_action.certificate_quantity
+            ):
+                certificates_to_transfer.append(gc_bundle)
+                continue
+
+        chlid_bundle_1, _child_bundle_2 = split_certificate_bundle(
+            gc_bundle,
+            certificates_to_split[idx],
+            write_session,
+            read_session,
+            esdb_client,
+        )
+        if chlid_bundle_1:
+            certificates_to_transfer.append(chlid_bundle_1)
+
+    return certificates_to_transfer
+
+
 def query_certificates(
```

Comment:
I think it would be easier to comprehend and give better separation of concerns to split this out into different end-points and services e.g.:

- get_certificates_by_device_timestamp_list
- get_certificates_issued_in_previous_week_by_device